### PR TITLE
Happy eyeballs handling when a connect attempt completes and fails before the second lookup completes and returns NODATA.

### PR DESCRIPTION
**Pull request by jhatala**
If the connection attempt has been under way for more than CONNECTION_ATTEMPT_DELAY_MS and the lookup that just completed gave us a new address to try, then stop that connection attempt and start a new connection attempt using the new address.

If the connection attempt has been under way for less than that, then do nothing for now.  Eventually, either the timeout will expire or the connection attempt will complete.

If the connection attempt is under way but the lookup has not provided us any new address to try, then do nothing for now, and wait for the connection attempt to complete.

Before this change, the early return when there are no new addresses would have been taken even when no connection attempt is under way.  Eventually the request would time out, even though it should have failed quickly.

### CODE REVIEW

File: lib/handler/connect.c [hunk 1]
```diff
@@ -416,9 +416,6 @@ static void on_getaddr(h2o_hostinfo_getaddr_req_t *getaddr_req, const char *errs
         }
     }
 
-    /* Try connecting if possible. */
-    if (self->server_addresses.used == self->server_addresses.size)
-        return;
```
Comment (kazuho) [inline lib/handler/connect.c:421]: Thank you for working on the fix. The problem statement sounds correct to me.

OTOH, I'm not sure if the proposed solution is correct.

Consider the following case:
1. ipv4 lookup succeeds
2. after `name_resolution_delay` milliseconds, we attempt to connect to that IPv4 address
3. while the connection attempt is ongoing, ipv6 lookup fails
In this case, we want the connection attempt to continue running. However, with the proposed solution it seems to me that we'd be closing the socket immediately.

I wonder if the correct fix is to move this if statement inside `if (self->sock != NULL) {` right below.

Comment (jhatala) [reply lib/handler/connect.c:421]: I understand, thank you very much for the explanation!

I pushed a new version with the two lines moved inside the condition on there being a pending connection attempt, and I added comments that (to me) explain the three cases that the code handles.

**Comment by jhatala** [comment]
(The ci failure is redis failing to start due to a port collision.)

<details>

```
2022-08-10T23:20:56.0336024Z   6.116686 25032:C 10 Aug 2022 23:20:43.016 # Redis version=6.0.16, bits=64, commit=00000000, modified=0, pid=25032, just started
2022-08-10T23:20:56.0336360Z   6.116699 25032:C 10 Aug 2022 23:20:43.016 # Configuration loaded
2022-08-10T23:20:56.0336821Z   6.117431 25032:M 10 Aug 2022 23:20:43.016 # You requested maxclients of 10000 requiring at least 10032 max file descriptors.
2022-08-10T23:20:56.0337406Z   6.117444 25032:M 10 Aug 2022 23:20:43.016 # Server can't set maximum open files to 10032 because of OS error: Operation not permitted.
2022-08-10T23:20:56.0338103Z   6.117455 25032:M 10 Aug 2022 23:20:43.016 # Current maximum open files is 1024. maxclients has been reduced to 992 to compensate for low ulimit. If you need higher maxclients increase 'ulimit -n'.
2022-08-10T23:20:56.0338599Z   6.117516 25032:M 10 Aug 2022 23:20:43.016 # Could not create server TCP listening socket *:45157: bind: Address already in use
2022-08-10T23:20:56.0338903Z   6.212167 done
2022-08-10T23:20:56.0339163Z   6.359195 spawning /home/ci/h2o... [INFO] raised RLIMIT_NOFILE to 1024
2022-08-10T23:20:56.0339506Z   6.375169 h2o server (pid:25059) is ready to serve requests with 1 threads
2022-08-10T23:20:56.0339894Z   6.375358 mruby raised: Connection Error (Kernel::H2O::Redis::ConnectionError)
2022-08-10T23:20:56.0340268Z   6.375537 mruby raised: Connection Error (Kernel::H2O::Redis::ConnectionError)
2022-08-10T23:20:56.0340552Z   6.414402 done
2022-08-10T23:20:56.0340991Z   6.426476 fetch-ocsp-response (using OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022))
2022-08-10T23:20:56.0341366Z   6.439199 failed to extract ocsp URI from /tmp/5mKAaBCEMT/cert.crt
2022-08-10T23:20:56.0341711Z   6.444411 [OCSP Stapling] disabled for certificate file:examples/h2o/server.crt
2022-08-10T23:20:56.0342063Z   6.458058             not ok 1
2022-08-10T23:20:56.0342485Z   6.458073             #   Failed test at t/50mruby-redis.t line 309.
2022-08-10T23:20:56.0371375Z   6.458076             #          got: '500'
2022-08-10T23:20:56.0371797Z   6.458078             #     expected: '200'
2022-08-10T23:20:56.0372010Z   6.458080             not ok 2
```

</details>

**Comment by kazuho** [comment]
Thank you for the changes. Looks perfect (including the comment!).

**Pull request merged**